contracts/src/v0.8/automation/upkeeps/UpkeepBalanceMonitor.sol

```diff
@@ -0,0 +1,232 @@
+
+  /// @param linkToken the Link token address
+  /// @param config the initial config for the contract
+  constructor(LinkTokenInterface linkToken, Config memory config) ConfirmedOwner(msg.sender) {
+    require(address(linkToken) != address(0));
+    LINK_TOKEN = linkToken;
+    setConfig(config);
+  }
+
+  // ================================================================
+  // |                      CORE FUNCTIONALITY                      |
+  // ================================================================
+
+  /// @notice Gets a list of upkeeps that are underfunded
+  /// @return needsFunding list of underfunded upkeepIDs
+  /// @return topUpAmounts amount to top up each upkeep
+  function getUnderfundedUpkeeps() public view returns (uint256[] memory, uint256[] memory) {
+    uint256 numUpkeeps = s_watchList.length;
+    uint256[] memory needsFunding = new uint256[](numUpkeeps);
+    uint256[] memory topUpAmounts = new uint256[](numUpkeeps);
+    Config memory config = s_config;
+    IAutomationRegistryConsumer registry = getRegistry();
+    uint256 availableFunds = LINK_TOKEN.balanceOf(address(this));
+    uint256 count;
+    uint256 upkeepID;
+    for (uint256 i = 0; i < numUpkeeps; i++) {
+      upkeepID = s_watchList[i];
+      uint96 upkeepBalance = registry.getBalance(upkeepID);
+      uint256 minBalance = uint256(registry.getMinBalance(upkeepID));
+      uint256 topUpThreshold = (minBalance * config.minPercentage) / 100;
+      uint256 topUpAmount = ((minBalance * config.targetPercentage) / 100) - upkeepBalance;
+      if (topUpAmount > config.maxTopUpAmount) {
+        topUpAmount = config.maxTopUpAmount;
+      }
+      if (upkeepBalance <= topUpThreshold && availableFunds >= topUpAmount) {
+        needsFunding[count] = upkeepID;
+        topUpAmounts[count] = topUpAmount;
+        count++;
+        availableFunds -= topUpAmount;
+      }
+      if (count == config.maxBatchSize) {
+        break;
+      }
+    }
+    if (count < numUpkeeps) {
+      assembly {
+        mstore(needsFunding, count)
+        mstore(topUpAmounts, count)
+      }
+    }
+    return (needsFunding, topUpAmounts);
+  }
+
+  /// @notice Called by the keeper/owner to send funds to underfunded upkeeps
+  /// @param upkeepIDs the list of upkeep ids to fund
+  /// @param topUpAmounts the list of amounts to fund each upkeep with
+  /// @dev We explicitly choose not to verify that input upkeepIDs are included in the watchlist. We also
+  /// explicity permit any amount to be sent via topUpAmounts; it does not have to meet the criteria
+  /// specified in getUnderfundedUpkeeps(). Here, we are relying on the security of automation's OCR to
+  /// secure the output of getUnderfundedUpkeeps() as the input to topUp(), and we are treating the owner
+  /// as a privileged user that can perform arbitrary top-ups to any upkeepID.
+  function topUp(uint256[] memory upkeepIDs, uint96[] memory topUpAmounts) public {
+    IAutomationForwarder forwarder = s_forwarder;
```

Comment:
I dont see the setForwarder function is invoked anywhere in this contract, how is forwarder set?